keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1561 @@
+
+Some applications solve the disruption problem by introducing validating admission webhooks.
+This has some drawbacks. The webhooks are not easily discoverable by cluster admins. And they can
+block evictions for other applications if they are misconfigured or misbehave. The eviction API is
+not intended to be extensible in this way. The webhook approach is therefore not recommended.
+
+Some drainers solve the node drain by depending on the kubectl logic, or by extending/rewriting it
+with additional rules and logic.
+
+As seen in the experience reports and GitHub issues ([Declarative Node Maintenance KEP](https://github.com/kubernetes/enhancements/pull/4213)),
+some admins solve their problems by simply ignoring PDBs which can cause unnecessary disruptions or
+data loss. Some solve this by playing with the application deployment, but have to understand that
+the application supports this.
+
+
+### PodDisruptionBudget Standing Issues
+
+For completeness here is a complete list of upstream open PDB issues. Most are relevant to this KEP.
+
+- [Mandatorliy specify how the application handle disruptions in the pod spec.](https://github.com/kubernetes/kubernetes/issues/124390)
+- [Treat liveness probe-based restarts as voluntary disruptions gated by PodDisruptionBudgets](https://github.com/kubernetes/kubernetes/issues/123204)
+- [Correct Keyword name for DisruptionsAllowed in PDB.](https://github.com/kubernetes/kubernetes/issues/121585)
+- [maxSurge for node draining or how to meet availability requirements when draining nodes by adding pods](https://github.com/kubernetes/kubernetes/issues/114877)
+- [Topology Aware Infrastructure Disruptions for Statefulsets](https://github.com/kubernetes/kubernetes/issues/114010)
+- [Support PDBs for DS](https://github.com/kubernetes/kubernetes/issues/108124)
+- [Allow scaling up to meet PDB constraints](https://github.com/kubernetes/kubernetes/issues/93476)
+- [Priority-based preemption can easily violate PDBs even when unnecessary due to multiple issues with the implementation](https://github.com/kubernetes/kubernetes/issues/91492)
+- [Disruption controller support configure workers' number](https://github.com/kubernetes/kubernetes/issues/82930)
+- [Cannot drain node with pod with more than one Pod Disruption Budget](https://github.com/kubernetes/kubernetes/issues/75957)
+- [Eviction should be able to request eviction asynchronously](https://github.com/kubernetes/kubernetes/issues/66811)
+- [Reasonable defaults with eviction and PodDisruptionBudget](https://github.com/kubernetes/kubernetes/issues/35318)
+
+### Goals
+
+- Introduce new `evacuation.coordination.k8s.io` API and evacuation controller.
+- Block eviction (Eviction API) of pods that support the evacuation process, unless the evacuation process fails.
+
+### Non-Goals
+
+- Implement evacuation capabilities in Kubernetes workloads (ReplicaSet, Deployment, etc.).
+- Synchronizing of the evacuation status to the pod status.
```

Comment:
does it mean that pod status will not reflect the pod is being evacuated? on a high level, how do we identify a pod is currently being evacuated?

Reply:
You would have to pair the pod to an evacuation. I think it would be a nice followup once we are set on the Evacuation API to add for example a pod condition to indicate that. This thread is also relevant: https://github.com/kubernetes/enhancements/pull/4565#discussion_r1780227788

Reply:
so what we monitor is the existing evacuations and check the evacuation statuses instead? if an evacuation that failed to pair with a pod is eventually garbage collected, will the pod ever know it was intended to be evacuated?

Reply:
> I think it would be a nice followup once we are set on the Evacuation API to add for example a pod condition to indicate that.

I think we can add this followup to the Beta graduation.

> so what we monitor is the existing evacuations and check the evacuation statuses instead?

yes, for now

> if an evacuation that failed to pair with a pod is eventually garbage collected, will the pod ever know it was intended to be evacuated?

The pairing is the responsibility of the Evacuator. Nevertheless, we should check as much as we can on admission.

Reply:
Thank you!

Reply:
Added to Beta graduation